### PR TITLE
group by view

### PR DESCRIPTION
**Pull request by tala-abuadas**

### CODE REVIEW

File: src/foam/u2/table/UnstyledTableView.js [hunk 1]
```diff
@@ -402,9 +404,10 @@ foam.CLASS({
       }
 
       //otherwise on adding new column creating new EditColumnsView, which is closed by default
-      if ( view.editColumnsEnabled )
-        var editColumnView = foam.u2.view.EditColumnsView.create({data:view}, this);
-
+      if ( view.editColumnsEnabled ) {
+        var editColumnView = this.EditColumnsView.create({ data: view, overlayView: { class: 'foam.u2.view.ColumnConfigPropView' }});
+        var groupByView = this.EditColumnsView.create({ data: view, overlayView: { class: 'foam.u2.view.GroupByView' }});
+      }
```
Comment (sarthak-marwaha) [inline src/foam/u2/table/UnstyledTableView.js]: Do we need these anymore?

File: src/foam/u2/view/GroupByView.js [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+ */
+
+foam.CLASS({
+  package: 'foam.u2.view',
+  name: 'GroupByView',
+  extends: 'foam.u2.view.ColumnConfigPropView',
+  requires: ['foam.u2.view.GroupByViewRow', 'foam.u2.UnstyledTableView','foam.u2.view.SubColumnSelectConfig'],
+  properties: [
+    'data',
+    'selectedCol',
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/GroupByView.js]: Why do we need this prop? We modify `this.data.groupBy` directly

File: src/foam/u2/view/GroupByView.js [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+              'padding-right': '8px'
+            })
+            .start()
+              .addClass(this.myClass('label'))
+              .callIfElse( this.theme , 
+                function() {
+                  this.add(self.slot(function( data$isPropertySelected ) { 
+                    var image;
+                    if ( ! this.data.isPropertySelected ) {
+                      image = self.theme.glyphs.folderFill.getDataUrl({ fill: self.theme.primary3});
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/GroupByView.js]:
```suggestion
                      image = self.theme.glyphs.folderFill.getDataUrl({ fill: self.theme.primary3 });
```
Same for the next one

**Review by sarthak-marwaha — CHANGES_REQUESTED** [review]

File: src/foam/u2/view/GroupByView.js [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.u2.view',
+  name: 'GroupByView',
+  extends: 'foam.u2.view.ColumnConfigPropView',
+  requires: ['foam.u2.view.GroupByViewRow', 'foam.u2.UnstyledTableView','foam.u2.view.SubColumnSelectConfig'],
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/GroupByView.js]:
```suggestion
  requires: [
  'foam.u2.view.GroupByViewRow',
  'foam.u2.UnstyledTableView',
  'foam.u2.view.SubColumnSelectConfig'
  ],
```

File: src/foam/u2/view/GroupByView.js [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            onSelectionChanged: this.onSelectionChanged.bind(this)
+          }));
+        }
+        return arr;
+      }
+    }
+  ]
+  ,
+  methods: [
+    function render() {
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/GroupByView.js]: Remove this render

File: src/foam/u2/view/GroupByView.js [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+  methods: [
+    function onSelect(draggableIndex, views) {
+      if ( this.selectCol ) {
+        this.selectCol.isPropertySelected = false;
+      }
+        var tc = views[draggableIndex].prop.rootProperty[0];
+        axiom = this.data.of.getAxiomByName(tc);
+        this.data.groupBy = axiom;
+        this.selectCol = views[draggableIndex].prop;
+        //views[draggableIndex].prop.isPropertySelected = ! views[draggableIndex].prop.isPropertySelected;
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/GroupByView.js]: remove comments

File: src/foam/u2/view/GroupByView.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    },
+    function onUnSelect(draggableIndex, views) {
+      this.data.groupBy = undefined;
+      this.selectCol = undefined;
+      //views[draggableIndex].prop.isPropertySelected = ! views[draggableIndex].prop.isPropertySelected;
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/GroupByView.js]: same here

File: src/foam/u2/view/GroupByView.js [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+          }));
+        }
+        return arr;
+      }
+    }
+  ]
+  ,
+  methods: [
+    function onSelect(draggableIndex, views) {
+      if ( this.selectCol ) {
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/GroupByView.js]: indent

**Review by sarthak-marwaha — COMMENTED** [review]
LGTM

**Review by sarthak-marwaha — APPROVED** [review]

File: src/foam/u2/view/EditColumnsView.js [hunk 9]
```diff
@@ -63,18 +63,16 @@ foam.CLASS({
       class: 'Boolean'
     },
     'parentId',
-    'columnConfigPropView'
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/EditColumnsView.js]: No trailing commas

File: src/foam/u2/view/ColumnConfigView.js [hunk 10]
```diff
@@ -457,10 +508,12 @@ foam.CLASS({
   package: 'foam.u2.view',
   name: 'ColumnViewHeader',
   extends: 'foam.u2.View',
-
-  requires: ['foam.u2.CheckBox'],
+  imports: ['theme'],
+  requires: [
+  'foam.u2.CheckBox',
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/ColumnConfigView.js]: indent

File: src/foam/u2/view/ColumnConfigView.js [hunk 11]
```diff
@@ -482,9 +535,13 @@ foam.CLASS({
     align-items: center;
     justify-content: start;
   }
+  ^selection-buttons + ^selection-buttons {
+  padding: 8px;
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/ColumnConfigView.js]: indent

File: src/foam/u2/view/ColumnConfigView.js [hunk 12]
```diff
@@ -509,11 +566,37 @@ foam.CLASS({
...
+                 .addClass(this.myClass('selection-buttons'))
+                 .call(function() {
+                   if ( self.data.level > 1 ) return;
+                   if ( self.theme ) {
+                     this.add(self.slot(function( data$isPropertyGrouped ) {
+                       var image;
+                       if ( self.data.isPropertyGrouped ) {
+                         image = self.theme.glyphs.folderFill.getDataUrl({ fill: self.theme.primary3 });
+                       } else {
+                         image = self.theme.glyphs.folderOutline.getDataUrl({ fill: self.theme.grey4 });
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/ColumnConfigView.js]: fill here should be `grey2` to match the checkbox outline

File: src/foam/u2/table/UnstyledTableView.js [hunk 13]
```diff
@@ -644,10 +642,9 @@ foam.CLASS({
         parent: this.__subContext__.createSubContext({ memento: this.currentMemento_ && this.currentMemento_.tail })
       }));
     },
-  ]
+  ],
```
Comment (sarthak-marwaha) [inline src/foam/u2/table/UnstyledTableView.js]: trailing commas

File: src/foam/u2/table/UnstyledTableView.js [hunk 14]
```diff
@@ -56,7 +56,7 @@ foam.CLASS({
     {
       type: 'Int',
       name: 'EDIT_COLUMNS_BUTTON_CONTAINER_WIDTH',
-      value: 60
+      value: 120
```
Comment (sarthak-marwaha) [inline src/foam/u2/table/UnstyledTableView.js]: Do we need to do this now that we have only one button?

File: src/foam/u2/table/TableView.js [hunk 15]
```diff
@@ -137,5 +137,11 @@
       height: auto;
       border-bottom: 2px solid /*%PRIMARY3%*/ #406DEA;
     }
+    ^th-editColumns {
```
Comment (sarthak-marwaha) [inline src/foam/u2/table/TableView.js]: Same here

File: src/foam/u2/view/ColumnConfigView.js [hunk 16]
```diff
@@ -384,6 +428,10 @@ foam.CLASS({
       name: 'onSelectionChangedParentFunction',
       documentation: 'parent\'s on this onSelectionChanged function'
     },
+    {
+      name: 'onGroupByChangedParentFunction',
+      documentation: 'parent\'s on this onSelectionChanged function'
```
Comment (sarthak-marwaha) [inline src/foam/u2/view/ColumnConfigView.js]: update documentation

File: src/foam/u2/table/UnstyledTableView.js (hunk 14 above)
Comment (tala-abuadas) [reply src/foam/u2/table/UnstyledTableView.js]: no

**Review by sarthak-marwaha — APPROVED** [review]

File: src/foam/u2/view/LazyScrollManager.js [hunk 17]
```diff
@@ -112,7 +112,7 @@
       name: 'scrollToIndex',
       postSet: function () { this.safeScroll(); }
     },
-    'currGroup_',
+    ['currGroup_', undefined],
```
Comment (kgrgreer) [inline src/foam/u2/view/LazyScrollManager.js]: What is this line needed for? undefined isn't a valid default value

**Pull request merged**